server/api/trend.go

```diff
@@ -0,0 +1,183 @@
+	HotReadFlow         uint64   `json:"hot_read_flow"`
+	HotReadRegionFlows  []uint64 `json:"hot_read_region_flows"`
+}
+
+type trendHistory struct {
+	StartTime int64               `json:"start"`
+	EndTime   int64               `json:"end"`
+	Entries   []trendHistoryEntry `json:"entries"`
+}
+
+type trendHistoryEntry struct {
+	From  uint64 `json:"from"`
+	To    uint64 `json:"to"`
+	Kind  string `json:"kind"`
+	Count int    `json:"count"`
+}
+
+type trendHandler struct {
+	*server.Handler
+	svr *server.Server
+	rd  *render.Render
+}
+
+func newTrendHandler(s *server.Server, rd *render.Render) *trendHandler {
+	return &trendHandler{
+		Handler: s.GetHandler(),
+		svr:     s,
+		rd:      rd,
+	}
+}
+
+func (h *trendHandler) Handle(w http.ResponseWriter, r *http.Request) {
+	var from time.Time
+	if fromStr := r.URL.Query()["from"]; len(fromStr) > 0 {
+		fromInt, err := strconv.ParseInt(fromStr[0], 10, 64)
+		if err != nil {
+			h.rd.JSON(w, http.StatusBadRequest, err.Error())
+			return
+		}
+		from = time.Unix(fromInt, 0)
+	}
+
+	stores, err := h.getTrendStores()
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	history, err := h.getTrendHistory(from)
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	trend := Trend{
+		Stores:  stores,
+		History: history,
+	}
+	h.rd.JSON(w, http.StatusOK, trend)
+}
+
+func (h *trendHandler) getTrendStores() ([]trendStore, error) {
+	maxStoreDownTime := h.svr.GetScheduleConfig().MaxStoreDownTime.Duration
+
+	var readStats, writeStats core.StoreHotRegionsStat
+	if hotRead := h.GetHotReadRegions(); hotRead != nil {
+		readStats = hotRead.AsLeader
+	}
+	if hotWrite := h.GetHotWriteRegions(); hotWrite != nil {
+		writeStats = hotWrite.AsPeer
+	}
+	stores, err := h.GetStores()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	trendStores := make([]trendStore, 0, len(stores))
+	for _, store := range stores {
+		info := newStoreInfo(store, maxStoreDownTime)
+		s := trendStore{
+			ID:              info.Store.GetId(),
+			Address:         info.Store.GetAddress(),
+			StateName:       info.Store.StateName,
+			Capacity:        uint64(info.Status.Capacity),
+			Available:       uint64(info.Status.Available),
+			RegionCount:     info.Status.RegionCount,
+			LeaderCount:     info.Status.LeaderCount,
+			StartTS:         info.Status.StartTS,
+			LastHeartbeatTS: info.Status.LastHeartbeatTS,
+			Uptime:          info.Status.Uptime,
+		}
+		s.HotReadFlow, s.HotReadRegionFlows = h.getStoreFlow(readStats, store.GetId())
+		s.HotWriteFlow, s.HotWriteRegionFlows = h.getStoreFlow(writeStats, store.GetId())
+		trendStores = append(trendStores, s)
+	}
+	return trendStores, nil
+}
+
+func (h *trendHandler) getStoreFlow(stats core.StoreHotRegionsStat, storeID uint64) (storeFlow uint64, regionFlows []uint64) {
+	if stats == nil {
+		return
+	}
+	if stat, ok := stats[storeID]; ok {
```

Comment:
Not thread-safe here by design?

Reply:
It's safe as each request are accessing different copy.